src/ooo_cpu.cc

```diff
@@ -475,13 +525,39 @@ void O3_CPU::decode_and_dispatch()
     if (DECODE_BUFFER.occupancy == 0)
         return;
 
+    std::size_t available_decode_bandwidth = DECODE_WIDTH;
+
     // dispatch DECODE_WIDTH instructions that have decoded into the ROB
-    uint32_t count_dispatches = 0;
-    while (count_dispatches < DECODE_WIDTH && DECODE_BUFFER.occupancy > 0 && ROB.occupancy < ROB.SIZE &&
+    while (available_decode_bandwidth > 0 && DECODE_BUFFER.occupancy > 0 && ROB.occupancy < ROB.SIZE &&
             (!warmup_complete[cpu] || ((DECODE_BUFFER.entry[DECODE_BUFFER.head].decoded) && (DECODE_BUFFER.entry[DECODE_BUFFER.head].event_cycle < current_core_cycle[cpu]))))
     {
+        ooo_model_instr &db_entry = DECODE_BUFFER.entry[DECODE_BUFFER.head];
+
+	// Search DIB to see if we need to add this instruction
+	dib_t::value_type &dib_hit_check_set = DIB[(db_entry.ip >> LOG2_BLOCK_SIZE) % DIB_SET];
+	auto way = std::find_if(dib_hit_check_set.begin(), dib_hit_check_set.end(), [db_entry](dib_entry_t x){ return x.valid && ((x.addr >> LOG2_BLOCK_SIZE) == (db_entry.ip >> LOG2_BLOCK_SIZE));});
+	if (way == dib_hit_check_set.end())
+	  {
+	    // we have a miss in the DIB, so we need to replace something
+	    // find victim in DIB
+	    dib_t::value_type &dib_set = DIB[db_entry.ip % DIB_SET];
+	    auto way = std::find_if_not(dib_set.begin(), dib_set.end(), [](dib_entry_t x){ return x.valid; }); // search for invalid
+	    if (way == dib_set.end())
+	      way = std::max_element(dib_set.begin(), dib_set.end(), [](dib_entry_t x, dib_entry_t y){ return x.lru < y.lru;}); // search for LRU
```

Comment:
These statements could all be combined into
```suggestion
	    dib_t::value_type &dib_set = DIB[db_entry.ip % DIB_SET];
	    auto way = std::max_element(dib_set.begin(), dib_set.end(), [](dib_entry_t x, dib_entry_t y){ return !y.valid || (x.valid && x.lru < y.lru;}); // search for LRU
```

This would remove an if-statement and be more concise.

Reply:
Honestly, I don't exactly understand this code.  I just copied/pasted it from elsewhere and I assume you wrote it correctly in the first instance.  After my changes are merged into develop, will you please take a closer look at this?

Reply:
I added a commit that does this. I admit, the syntax is a bit clunky, but the code just searches through to find a victim in the DIB, where invalid ways are always the most eligible candidates.